### PR TITLE
populate WCS keywords in Level1 files

### LINKED ISSUE #2932: set_telescope_pointing should populate all FITS WCS keywords
Issue [JP-475](https://jira.stsci.edu/browse/JP-475) was created by Howard Bushouse:

Currently the FITS WCS keywords in level-1b products are partially populated by SDP code and others by the SCSB set_telescope_pointing script. SDP has agreed to move the population of all WCS keywords in level-1b's  to the set_telescope_pointing script, so that it's all contained in one place.

Previously documented in github issue [#2597|https://github.com/spacetelescope/jwst/issues/2597]

### PR DESCRIPTION
**Pull request by nden**
This is based and should be merged after #3065.

Close #2932 and #2262.

TODO:

[X] Test with the SIAF database

[] Edit  [this page](https://outerspace.stsci.edu/pages/viewpage.action?spaceKey=JWSTPWG&title=Required+Header+Keywords+for+Test+Exposures) to include the changes in this PR and pass it to the INS teams

### CODE REVIEW

**Comment by nden** [comment]
This is now ready for final review.
Before merging it should be tested on the D string if possible.

File: jwst/lib/exposure_types.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+                                         IMAGE2_NONSCIENCE_EXP_TYPES,
+                                         SPEC2_SCIENCE_EXP_TYPES)
+
+IMAGING_TYPES = set(tuple(ACQ_EXP_TYPES) + tuple(IMAGE2_SCIENCE_EXP_TYPES)
+                    + tuple(IMAGE2_NONSCIENCE_EXP_TYPES) +
+                    ('fgs_image', 'fgs_focus'))
+
+SPEC_TYPES = SPEC2_SCIENCE_EXP_TYPES
+
+# JWST Exposures that are Fine Guidance exposures that actually
```
Comment (hbushouse) [inline jwst/lib/exposure_types.py]: A really minor nit, but "Fine Guidance" is actually the name of one of the FGS guiding modes (ID, Acq, Track, Fine Guide), so a more appropriate wording for any type of guiding mode exposures would be "FGS guide star exposures". And technically they aren't even "exposures"; they're a continuous data stream that lasts as long as a given type of guiding activity takes to complete (e.g. Track and Fine Guide data could span time covering many exposures), but that's an even more minor nit that we won't worry about here.

File: jwst/lib/set_telescope_pointing.py [hunk 2]
```diff
@@ -237,13 +248,20 @@ def update_wcs(model, default_pa_v3=0., siaf_path=None, engdb_url=None,
         exp_type = model.meta.exposure.type.lower()
     except AttributeError:
         exp_type = None
+    aperture_name = model.meta.aperture.name.upper()
+    if aperture_name is not "UNKNOWN":
+        logger.info("Updating WCS for aperture {}".format(aperture_name))
+        useafter = model.meta.observation.date
+        siaf = _read_wcs_from_siaf(aperture_name, useafter, siaf_path)
+        populate_model_from_siaf(model, siaf)
+
```
Comment (hbushouse) [inline jwst/lib/set_telescope_pointing.py]: Should we have an "else" clause here that issues a warning (or at least info) when aperture_name IS unknown?

File: jwst/lib/set_telescope_pointing.py [hunk 3]
```diff
@@ -413,37 +425,33 @@ def update_wcs_from_telem(
...
-     model.meta.wcsinfo.pc2_1,
-     model.meta.wcsinfo.pc2_2) = calc_rotation_matrix(
-         wcsinfo.pa * D2R, vparity=siaf.vparity
-     )
-
-    wcsaxes = model.meta.wcsinfo.wcsaxes
-    if wcsaxes is None:
-        wcsaxes = 2
-    model.meta.wcsinfo.wcsaxes = max(2, wcsaxes)
+    if model.meta.exposure.type.lower() in TYPES_TO_UPDATE:
```
Comment (hbushouse) [inline jwst/lib/set_telescope_pointing.py]: TYPES_TO_UPDATE includes the FGS guide star modes. Is there a separate branch somewhere in the middle here (that isn't visible in the diff) that does the special WCS computation for those data sets? They are handled differently than normal science exposures.

File: jwst/lib/set_telescope_pointing.py [hunk 4]
```diff
@@ -1110,54 +1104,91 @@ def _roll_angle_from_matrix(matrix, v2, v3):
     return new_roll
 
 
-def _get_vertices_idl(aperture_name, useafter, prd_db_filepath=None):
+def _read_wcs_from_siaf(aperture_name, useafter, prd_db_filepath=None):
+    """
+    Read WCS keywords from the SIAF database file.
```
Comment (hbushouse) [inline jwst/lib/set_telescope_pointing.py]: I'm not sure I would characterize this as reading "WCS keywords" from the SIAF. We're getting values from the SIAF that are then used to populate FITS WCS keywords in the science data. There aren't any actual WCS keywords in the SIAF (but it's only a doc string, not working code, so ...)

File: jwst/lib/set_telescope_pointing.py [hunk 5]
```diff
@@ -1110,54 +1104,91 @@ def _roll_angle_from_matrix(matrix, v2, v3):
...
+
+    Given an ``APERTURE_NAME`` and a ``USEAFTER`` dat query the SIAF database
```
Comment (hbushouse) [inline jwst/lib/set_telescope_pointing.py]: "dat" should be "date"

File: jwst/lib/set_telescope_pointing.py [hunk 6]
```diff
@@ -1110,54 +1104,91 @@ def _roll_angle_from_matrix(matrix, v2, v3):
...
-    Adds CTYPE3 and CUNIT3 for spectral observations.
-    This is temporary, may be moved to SDP proccessing later.
-    """
-    wcsaxes = model.meta.wcsinfo.wcsaxes
-    if wcsaxes is not None and wcsaxes == 3:
-        model.meta.wcsinfo.ctype3 = "WAVE"
-        model.meta.wcsinfo.cunit3 = "um"
+    if RESULT:
+        values = list(RESULT.values())[0]
+        # This call populates the SIAF tuple with the values fomr the database.
```
Comment (hbushouse) [inline jwst/lib/set_telescope_pointing.py]: "fomr" should be "from"

File: jwst/lib/set_telescope_pointing.py [hunk 7]
```diff
@@ -1110,54 +1104,91 @@ def _roll_angle_from_matrix(matrix, v2, v3):
...
+        # The last 8 values returned from the database are the vertices.
+        # They are wrapped in a list and assigned tp SIAF.vertices_idl.
```
Comment (hbushouse) [inline jwst/lib/set_telescope_pointing.py]: "tp" should be "to"

File: jwst/lib/set_telescope_pointing.py [hunk 8]
```diff
@@ -1364,6 +1395,36 @@ def all_pointings(mnemonics):
...
+        Input data as Level1bModel.
+    siaf : namedtuple
+        The WCS keywords read in from the SIAF.
+    """
+    # If not an imaging mode, update the pointing only.
+    model.meta.wcsinfo.v2_ref = siaf.v2_ref
+    model.meta.wcsinfo.v3_ref = siaf.v3_ref
+    model.meta.wcsinfo.v3yangle = siaf.v3yangle
+    model.meta.wcsinfo.vparity = siaf.vparity
+    if model.meta.exposure.type.lower() in TYPES_TO_UPDATE:
```
Comment (hbushouse) [inline jwst/lib/set_telescope_pointing.py]: Again, is this OK for the guide star mode data? The guide modes don't have aperture entries in the SIAF and hence siaf.crpix1 and siaf.crpix2 won't have values that can be extracted directly from the SIAF. They require alternate computation (I believe we simply set them to the middle of the image data array?).

**Comment by jdavies-st** [comment]
Would it be worthwhile running the regression tests against this branch on Jenkins?  I can kick off a run easily.

**Comment by hbushouse** [comment]
I'm not sure we have any reg tests that run `set_telescope_pointing`, due to lack of access to the SIAF in the test environment.

**Comment by stscieisenhamer** [comment]
There are a number of tests, a number of which refer to the siaf database.  The existing tests are "unit" tests, so no need to run through regression.

File: jwst/lib/set_telescope_pointing.py (hunk 3 above)
Comment (nden) [reply jwst/lib/set_telescope_pointing.py]: I removed the guider modes from TYPES_TO_UPDATE

File: jwst/lib/set_telescope_pointing.py (hunk 2 above)
Comment (nden) [reply jwst/lib/set_telescope_pointing.py]: Done

**Comment by nden** [comment]
This still needs a review check in order to be merged.

File: jwst/lib/set_telescope_pointing.py [hunk 9]
```diff
@@ -1364,6 +1399,36 @@ def all_pointings(mnemonics):
...
+        # For imaging modes update the pointing and
+        # the FITS WCS keywords.
+        model.meta.wcsinfo.ctype1 = 'RA--TAN'
+        model.meta.wcsinfo.ctype2 = 'DEC-TAN'
+        model.meta.wcsinfo.wcsaxes = 2
+        model.meta.wcsinfo.cunit1 = "deg"
+        model.meta.wcsinfo.cunit2 = "deg"
+        model.meta.wcsinfo.crpix1 = siaf.crpix1
+        model.meta.wcsinfo.crpix2 = siaf.crpix2
+        model.meta.coordinates.reference_frmae = "ICRS"
```
Comment (hbushouse) [inline jwst/lib/set_telescope_pointing.py]: "reference_frmae" needs to be "reference_frame"

Comment (nden) [reply jwst/lib/set_telescope_pointing.py]: Good catch! Fixed now.

Comment (hbushouse) [reply jwst/lib/set_telescope_pointing.py]: Sometimes I'm amazed how many little things like that can jump out just from a quick scan with a different set of eyes. 👍

**Pull request merged**